stdlib/public/Concurrency/Actor.cpp

```diff
@@ -314,39 +315,83 @@ bool _task_serialExecutor_isSameExclusiveExecutionContext(
     const SerialExecutorWitnessTable *wtable);
 
 SWIFT_CC(swift)
-static bool swift_task_isCurrentExecutorImpl(SerialExecutorRef executor) {
+static bool swift_task_isCurrentExecutorImpl(SerialExecutorRef expectedExecutor) {
   auto current = ExecutorTrackingInfo::current();
 
   if (!current) {
     // TODO(ktoso): checking the "is main thread" is not correct, main executor can be not main thread, relates to rdar://106188692
-    return executor.isMainExecutor() && isExecutingOnMainThread();
+    if (expectedExecutor.isMainExecutor() && isExecutingOnMainThread()) {
+      return true;
+    } else {
+      swift_Concurrency_fatalError(
+          0,
+          "Incorrect executor assumption, expected MainActor.executor (i.e. 'main thread')");
+    }
   }
 
-  auto currentExecutor = current->getActiveExecutor();
-  if (currentExecutor == executor) {
+  SerialExecutorRef currentExecutor = current->getActiveExecutor();
+  if (currentExecutor == expectedExecutor) {
+    return true;
+  }
+  if (currentExecutor.isMainExecutor() && expectedExecutor.isMainExecutor()) {
     return true;
   }
 
-  if (executor.isComplexEquality()) {
+  if (expectedExecutor.isDefaultActor()) {
+    swift_Concurrency_fatalError(0, "Incorrect actor executor assumption");
+    return false;
+  }
+
+  if (expectedExecutor.isMainExecutor() && !currentExecutor.isMainExecutor()) {
```

Comment:
Also here - We need to consider dispatch serial queue targeting the main queue, right?

Reply:
Looping back to this, sorry to keep you waiting.

Agreed in pricinple, but we should carefully roll out this change -- if we did this now, we would get much worse error messages. The current message includes `"expected MainActor executor"` which the _default_ implementation that'd SerialDispatchQueue would bene up using would not have.

We need dispatch to implement these APIs with good error messages "you expected main but was something else" I think?

If we think this doesn't matter I could already make this call checkIsolated though, but I think we should stagger the rollout ^